xl2times/transforms.py

```diff
@@ -1266,44 +1206,93 @@ def _populate_calculated_defaults(df: DataFrame, model: TimesModel):
+                ),
+                axis=1,
+            )
 
-        return replace(table, dataframe=df)
+            df.loc[i, "year"] = df[i].apply(
+                lambda row: (
+                    int(row["year"].split("-")[0])
+                    if "-" in str(row["year"])
+                    else (row["year"] if row["year"] != "" else pd.NA)
+                ),
+                axis=1,
+            )
+        if tag not in exclude_tags:
+            for colname in df.columns:
+                # TODO make this more declarative
+                # Apply default values to missing cells
+
+                if colname in mapping_to_defaults and "attribute" in df.columns:
+                    isna = df[colname].isna()
+                    if any(isna):
+                        key = mapping_to_defaults[colname]
+                        for value in config.veda_attr_defaults[key].keys():
+                            df.loc[
+                                isna
+                                & df["attribute"].isin(
+                                    config.veda_attr_defaults[key][value]
+                                ),
+                                colname,
+                            ] = value
+                elif colname == "year":
+                    df.loc[df[colname].isna(), [colname]] = model.start_year
+                elif colname == "currency":
+                    df.loc[df[colname].isna(), [colname]] = currency
+
+        # Do some additional processing for the tables
+        if tag in attribute_tags:
+            # TODO: should we have a global list of column name -> type?
+            if "year" in df.columns:
+                df["year"] = pd.to_numeric(df["year"], errors="coerce")
+            # Populate commodity and other_indexes based on defaults
+            for col in ("commodity", "other_indexes", "cg"):
+                _populate_defaults(tag, df, col, config, "original_attr")
 
-    return [apply_fixups_table(table) for table in tables]
+        tables[tag] = df
+
+    return tables
 
 
-def generate_commodity_groups(
+def include_cgs_in_topology(
     config: Config,
-    tables: list[EmbeddedXlTable],
+    tables: dict[str, DataFrame],
     model: TimesModel,
-) -> list[EmbeddedXlTable]:
-    """Generate commodity groups."""
-    process_tables = [t for t in tables if t.tag == Tag.fi_process]
-
+) -> dict[str, DataFrame]:
+    """Include commodity groups in model topology."""
     # Veda determines default PCG based on predetermined order and presence of OUT/IN commodity
     columns = ["region", "process", "primarycg"]
-    reg_prc_pcg = pd.DataFrame(columns=columns)
-    for process_table in process_tables:
-        df = process_table.dataframe[columns]
-        reg_prc_pcg = pd.concat([reg_prc_pcg, df])
-    reg_prc_pcg.drop_duplicates(keep="first", inplace=True)
+    reg_prc_pcg = tables[Tag.fi_process][columns].copy()
+    reg_prc_pcg.drop_duplicates(keep="last", inplace=True)
```

Comment:
```suggestion
    reg_prc_pcg = tables[Tag.fi_process][columns]
    reg_prc_pcg.drop_duplicates(keep="last")
```
This should be equivalent, no? And I think pandas recommends against inplace operators, as it's less efficient.